### PR TITLE
Add analytic event tracking to auction registration route

### PR DESCRIPTION
**Pull request by dblandin**
Jira: https://artsyproduct.atlassian.net/browse/AUCT-584

Event Registry: https://docs.google.com/spreadsheets/d/17oqFli15x1SmNBu_-fPZhFifMw2BNSPJe9VdOgs1aKs/edit#gid=0

## Local QA Gotcha

You'll need to `yarn add react-fast-compare@2.0.4` in Force for [reasons](https://artsy.slack.com/archives/C2TQ4PT8R/p1567089484035300)

### CODE REVIEW

**Comment by dleve123** [comment]
We think that this PR fires the same number and quality of events that currently exists on production

File: src/Apps/Auction/Components/RegistrationForm.tsx [hunk 1]
```diff
@@ -165,7 +167,13 @@ const InnerForm: React.FC<FormikProps<FormValues>> = props => {
         )}
       </Flex>
 
-      <Button mt={1} size="large" width="100%" loading={isSubmitting}>
+      <Button
+        mt={1}
+        size="large"
+        width="100%"
+        loading={isSubmitting}
+        {...{ type: "submit" } as any}
```
Comment (erikdstock) [inline src/Apps/Auction/Components/RegistrationForm.tsx]: We need `type="submit"` according to a console warning but it is not available on a palette button 🙃

File: src/Apps/Auction/Components/RegistrationForm.tsx [hunk 2]
```diff
@@ -186,17 +194,54 @@ const validationSchema = Yup.object().shape({
   ),
 })
 
+type TrackErrors = (errors: string[]) => void
```
Comment (erikdstock) [inline src/Apps/Auction/Components/RegistrationForm.tsx]: We pass our form submission error handler in from above where tracking is set up. This is its proptype.

File: src/Apps/Auction/Components/RegistrationForm.tsx [hunk 3]
```diff
@@ -186,17 +195,41 @@ const validationSchema = Yup.object().shape({
...
+    }
+  }
+  React.useEffect(effect, [formikProps.isSubmitting])
+
+  return null
+}
+
 export interface RegistrationFormProps
   extends ReactStripeElements.InjectedStripeProps {
+  tracking?: TrackingProp
```
[the remark this answers is not in the document]
Comment (erikdstock) [reply src/Apps/Auction/Components/RegistrationForm.tsx]: 🔪

File: src/Apps/Auction/Components/__stories__/RegistrationForm.story.tsx [hunk 4]
```diff
@@ -1,17 +1,23 @@
...
 import { storiesOf } from "storybook/storiesOf"
 import { Section } from "Utils/Section"
 import { StripeWrappedRegistrationForm } from "../RegistrationForm"
 
+const TrackedLayout = track()(Section)
+
 storiesOf("Apps/Auction/Components", module).add("RegistrationForm", () => {
   return (
-    <Section>
+    <TrackedLayout>
```
Comment (erikdstock) [inline src/Apps/Auction/Components/__stories__/RegistrationForm.story.tsx]: May not be needed any more- revisiting.

File: src/Apps/Auction/Routes/Register/index.tsx [hunk 5]
```diff
@@ -13,27 +7,69 @@ import {
...
+      ...commonProperties,
+    })
+  }
+
   function createBidder() {
     return new Promise(async (resolve, reject) => {
       commitMutation<RegisterCreateBidderMutation>(relay.environment, {
-        onCompleted: (data, errors) => {
-          resolve()
+        onCompleted: data => {
```
Comment (erikdstock) [inline src/Apps/Auction/Routes/Register/index.tsx]: Do we need to be handling the `(data, errors)` case rather than resolving unconditionally? Where does this second errors parameter come from rather than calling the `onError` parameter below?

File: src/Apps/Auction/Routes/Register/index.tsx [hunk 6]
```diff
@@ -109,16 +148,28 @@ export const RegisterRoute: React.FC<RegisterProps> = props => {
...
       .catch(error => {
         logger.error(error)
+        let errorMessages: string[]
+        if (Array.isArray(error)) {
+          errorMessages = error.map(e => e.message)
+        } else if (typeof error === "string") {
+          errorMessages = [error]
+        } else if (error.message) {
+          errorMessages = [error.message]
+        }
```
Comment (erikdstock) [inline src/Apps/Auction/Routes/Register/index.tsx]: In all cases we've tested the final `error.message` case applied here (though at least a `PayloadError[]` is a possible outcome), so this is just being pessimistic.

File: src/Apps/Auction/Routes/__tests__/RegisterRoute.test.tsx [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+      }
+      me {
+        ...Register_me
+      }
+    }
+  `,
+})
+
+xdescribe("Auction Registration App", () => {
+  // const getWrapper = (props: any) => {
```
Comment (erikdstock) [inline src/Apps/Auction/Routes/__tests__/RegisterRoute.test.tsx]: We haven't gotten _any_ of these tests working yet, so maybe we should not keep this file around.

File: src/Apps/Auction/Routes/__tests__/RegisterRouteAnalytics.test.tsx [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+    pending()
+  })
+
+  it("tracks form submission success", async () => {
+    pending()
+    // const wrapper = await renderRelayTree({
+    //   Component: (props: any) => (
+    //     <div>
+    //       <RegisterRouteFragmentContainer {...props} />
+    //     </div>
```
Comment (erikdstock) [inline src/Apps/Auction/Routes/__tests__/RegisterRouteAnalytics.test.tsx]: It's currently not easy to submit the form and mock out the entire success due to dependence on relay.

File: src/Apps/__tests__/Fixtures/Auction/Routes/Register/index.ts [hunk 9]
```diff
@@ -1,23 +1,44 @@
...
+import { routes_RegisterQueryResponse } from "__generated__/routes_RegisterQuery.graphql"
 
-export interface RegisterQueryResponse {
-  sale: redirects_sale
-  me: redirects_me
+type Omit<T, K> = Pick<T, Exclude<keyof T, K>>
+type DeFraged<T> = Omit<T, " $refType" | " $fragmentRefs">
+
+export interface DeFragedRegisterQueryResponse {
+  sale: DeFraged<routes_RegisterQueryResponse["sale"] & Register_sale>
```
Comment (erikdstock) [inline src/Apps/__tests__/Fixtures/Auction/Routes/Register/index.ts]: Our effort to remove confusing relay-specific attributes from these objects. If it's overkill, we will remove.

File: src/Apps/Auction/Components/__stories__/RegistrationForm.story.tsx (hunk 4 above)
Comment (erikdstock) [reply src/Apps/Auction/Components/__stories__/RegistrationForm.story.tsx]: 🔪

File: src/Apps/Auction/Routes/__tests__/RegisterRouteAnalytics.test.tsx [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+  it("tracks form submission failure when create bidder mutation fails", () => {
+    pending()
+  })
+
+  it("tracks form submission failure when create credit card Stripe API call fails", () => {
+    pending()
+  })
+
+  it("tracks form submission success", async () => {
+    pending()
```
Comment (erikdstock) [inline src/Apps/Auction/Routes/__tests__/RegisterRouteAnalytics.test.tsx]: It's currently not easy to submit the form and mock out the entire success or strange cases with the stripe api due to dependence on relay.

File: src/Apps/Auction/Components/RegistrationForm.tsx (hunk 1 above)
Comment (zephraph) [reply src/Apps/Auction/Components/RegistrationForm.tsx]: PRs welcome 😉

**Comment by dleve123** [comment]
I had myself as a reviewer to get Peril to stop it, assigned @zephraph as Erik and I were pairing on this.

File: src/Apps/Auction/Routes/Register/index.tsx (hunk 5 above)
Comment (zephraph) [reply src/Apps/Auction/Routes/Register/index.tsx]: `errors` here is different than `onError`. The `errors` parameter here are any potential errors that are returned from metaphysics. 

`onError` is an error that occurred with the request. I.E. the call to metaphysics likely failed.

File: src/Apps/Auction/Components/RegistrationForm.tsx (hunk 1 above)
Comment (erikdstock) [reply src/Apps/Auction/Components/RegistrationForm.tsx]: My understanding has been that we don't put html-specific attrs on palette components- what is the right way to do this?

Comment (zephraph) [reply src/Apps/Auction/Components/RegistrationForm.tsx]: That's not a strict constraint. We just don't want to add HTML props that flow down to react native.

Comment (zephraph) [reply src/Apps/Auction/Components/RegistrationForm.tsx]: Let's not worry with this right now. It's fine for this.

File: src/Apps/Auction/Routes/__tests__/RegisterRouteAnalytics.test.tsx (hunk 10 above)
Comment (zephraph) [reply src/Apps/Auction/Routes/__tests__/RegisterRouteAnalytics.test.tsx]: What are these `pending()` function calls? I don't see a reference to that anywhere.

File: src/Apps/Auction/__tests__/redirects.test.ts [hunk 11]
```diff
@@ -25,23 +25,23 @@ describe("Auction/redirects", () => {
...
-  ): RegisterQueryResponse => ({
+    data: DeFragedRegisterQueryResponse
+  ): DeFragedRegisterQueryResponse => ({
     sale: data.sale,
     me: data.me,
   })
 
   it("does not redirect if a sale is found", async () => {
-    const { redirect } = await render(
+    const result = await render(
```
Comment (zephraph) [inline src/Apps/Auction/__tests__/redirects.test.ts]: Why the change?

File: src/Apps/Auction/__tests__/redirects.test.ts [hunk 12]
```diff
@@ -25,23 +25,23 @@ describe("Auction/redirects", () => {
...
       mockResolver(Fixture)
     )
-
+    const { redirect } = result
     expect(redirect).toBeUndefined
   })
 
   it("redirects to the auction registration modal if the user has a qualified credit card", async () => {
-    const { redirect } = await render(
+    const result = await render(
```
Comment (zephraph) [inline src/Apps/Auction/__tests__/redirects.test.ts]: This one too. It's not evident why the destructuring was removed. Was it for debugging purposes?

File: src/Apps/Auction/__tests__/redirects.test.ts (hunk 11 above)
Comment (erikdstock) [reply src/Apps/Auction/__tests__/redirects.test.ts]: Artefact of debugging when we broke the relay found router.

File: src/Apps/Auction/__tests__/redirects.test.ts (hunk 12 above)
Comment (erikdstock) [reply src/Apps/Auction/__tests__/redirects.test.ts]: correct.

File: src/Apps/__tests__/Fixtures/Auction/Routes/Register/index.ts [hunk 13]
```diff
@@ -1,23 +1,44 @@
-import { redirects_me } from "__generated__/redirects_me.graphql"
-import { redirects_sale } from "__generated__/redirects_sale.graphql"
+import { Register_me } from "__generated__/Register_me.graphql"
+import { Register_sale } from "__generated__/Register_sale.graphql"
+import { routes_RegisterQueryResponse } from "__generated__/routes_RegisterQuery.graphql"
 
-export interface RegisterQueryResponse {
-  sale: redirects_sale
-  me: redirects_me
+type Omit<T, K> = Pick<T, Exclude<keyof T, K>>
```
Comment (zephraph) [inline src/Apps/__tests__/Fixtures/Auction/Routes/Register/index.ts]: Omit is a first class member of TS now, you shouldn't need to redefine it. 

https://www.typescriptlang.org/docs/handbook/utility-types.html#omittk

Comment (erikdstock) [reply src/Apps/__tests__/Fixtures/Auction/Routes/Register/index.ts]: I'm pretty sure we do need it unless we bump the ts version; it was working fine because we hadn't set pinned it to the editor version in `code`, but then it failed in type-checking

File: src/Apps/Auction/__tests__/redirects.test.ts (hunk 11 above)
Comment (erikdstock) [reply src/Apps/Auction/__tests__/redirects.test.ts]: (I can fix though)

File: src/Apps/__tests__/Fixtures/Auction/Routes/Register/index.ts (hunk 13 above)
Comment (zephraph) [reply src/Apps/__tests__/Fixtures/Auction/Routes/Register/index.ts]: All right, we can leave it for now. I'll look into bumping the version.

File: src/Apps/Auction/Components/RegistrationForm.tsx [hunk 14]
```diff
@@ -186,17 +192,53 @@ const validationSchema = Yup.object().shape({
...
+    if (
+      formikProps.submitCount > 0 &&
+      !formikProps.isSubmitting &&
+      !formikProps.isValid
+    ) {
+      cb(Object.values(formikProps.errors))
+      formikProps.setSubmitting(false)
+    }
+  }
+  React.useEffect(effect, [formikProps.submitCount, formikProps.isSubmitting])
```
Comment (zephraph) [inline src/Apps/Auction/Components/RegistrationForm.tsx]:
```suggestion
  React.useEffect(effect, [formikProps.submitCount, formikProps.isSubmitting, formikProps.isValid, formikProps.errors])
```

You need to track all props that are used within the effect.

File: src/Apps/Auction/Routes/__tests__/RegisterRouteAnalytics.test.tsx (hunk 10 above)
Comment (zephraph) [reply src/Apps/Auction/Routes/__tests__/RegisterRouteAnalytics.test.tsx]: If these tests just aren't implemented yet, I recommend using `it.todo(...` instead.

File: src/Apps/Auction/Components/RegistrationForm.tsx (hunk 14 above)
Comment (erikdstock) [reply src/Apps/Auction/Components/RegistrationForm.tsx]: I think this would actually introduce a bug- i know the general way of thinking about it is to track all parts as dependencies, but the purpose of this effect is to run only after failed form submission. Validation runs every time, so by letting `isValid` or `errors` be dependencies we this will change the behavior of the effect. This approach was taken from the formik issue linked in the doc comment and pr comment above & I actually commented on that issue to clarify that my understanding was right (still waiting to hear back).

**Comment by artsyit** [comment]
:rocket: PR was released in v17.9.0 :rocket:
